### ISSUE TITLE
Sample projects currently return 401 on requests for /userinfo

### ISSUE THREAD

**Issue by aheitzmann** [spring-projects/spring-authorization-server#716]
**Describe the bug**
The sample projects `default-authorizationserver` and `custom-consent-authorization-server` currently reject valid requests to the `/userinfo` endpoint.

**To Reproduce**
We can reproduce the issue with a couple of small additions to the existing `messages-client` sample application:

1) Add the following blocks to samples/messages-client/src/main/resources/templates/index.html,
- in the "list-group" `<ul>`:
```                
<li class="list-group-item">
  <a href="/userinfo"><span style="font-size:medium">Show User Info</span></a>
</li>
```
-  just above `<div th:if="${messages}" class="panel-footer">`:
```
<div th:if="${userInfo}" class="panel-footer">
  <h4>User Info</h4>
  <p th:text="${userInfo}">userinfo</p>
</div>
```
2) Add the following method to samples/messages-client/src/main/java/sample/web/AuthorizationController.java
```
	@GetMapping("/userinfo")
	public String getUserInfo(Model model, @RegisteredOAuth2AuthorizedClient("messaging-client-oidc")
			OAuth2AuthorizedClient authorizedClient) {

		String userInfo = this.webClient
				.get()
				.uri("http://127.0.0.1:9000/userinfo")
				.attributes(oauth2AuthorizedClient(authorizedClient))
				.retrieve()
				.bodyToMono(String.class)
				.block();
		model.addAttribute("userInfo", userInfo);
		return "index";
	}
```
3) Run the `default-authorizationserver` and `messages-client` sample applications
4) Navigate to http://localhost:8080/userinfo and log in with username `user1` and password `password`
This endpoint will make a request to the /userinfo endpoint of the authorization server using the user's access token as a bearer token and if it succeeds the response json will be displayed on the page. Set a breakpoint in the debugger to inspect the error.

**Expected behavior**
The page should show 
```
User Info
{"sub":"user1"}
```
under the Show User Info link.

**Comment by aheitzmann** [comment]
Some additional notes:
- https://github.com/spring-projects/spring-authorization-server/issues/398 removed `.oauth2ResourceServer(OAuth2ResourceServerConfigurer::jwt)` from the `HttpSecurity` configuration in `OAuth2AuthorizationServerConfiguration#applyDefaultSecurity`. This leaves the default authorization server configuration without any Filter or AuthenticationProvider that knows how to process bearer tokens.
- The samples can be easily fixed by adding
```
http.
  ...
  .oauth2ResourceServer(oauth2Configurer -> oauth2Configurer.jwt(jwtConfigurer -> jwtConfigurer.decoder(OAuth2AuthorizationServerConfiguration.jwtDecoder(jwkSource))))
  ...
```
to their existing HttpSecurity configurations in their respective `AuthorizationServerConfig#authorizationServerSecurityFilterChain` methods. However, it seems like since /userinfo is a standard part of the OIDC spec, enabling its use should not require such custom modifications.

**Comment by sjohnr** [comment]
@aheitzmann thanks for reaching out, and welcome to the project! See [this answer](https://stackoverflow.com/a/71341726/15835039) which states:

> Take a look at [OidcUserInfoTests](https://github.com/spring-projects/spring-authorization-server/blob/a661e1cdb7a77e3f2a1a0d958f8949896af66c6f/oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OidcUserInfoTests.java#L328), which has a configuration for demonstrating how to enable the use of user info endpoint.
>
> It requires the use of `.oauth2ResourceServer(OAuth2ResourceServerConfigurer::jwt)` and a `JwtDecoder` `@Bean` to allow the JWT access token to be validated.

Which is to say that you can enable resource server support and provide a `JwtDecoder`, and that is all that is required to enable the use of this endpoint.

> However, it seems like since /userinfo is a standard part of the OIDC spec, enabling its use should not require such custom modifications.

It is enabled by default, but it is not exposed by default which is by design because it is an optional endpoint in the specification. I'm going to close this issue for now, but if you have any additional thoughts on this feel free to add them.

**Comment by sjohnr** [comment]
Hi @aheitzmann, thanks for your response! You make good points and I appreciate you providing them.

> To my knowledge there is currently no documentation about how to set up or configure spring-authorization-server.

We are actively working on reference documentation as we speak that will be released later this month! You can see a preview [here](https://docs.spring.io/spring-authorization-server/docs/0.3.0-SNAPSHOT/reference/html/index.html). It will contain information about how to configure the userinfo endpoint, though I will be personally writing that section next week and it's not done yet. I am also in the process of writing a How-to guide on enabling and customizing the userinfo endpoint. You can see an early preview [here](https://github.com/sjohnr/spring-authorization-server/blob/gh-537-how-to-customize-user-info/docs/src/docs/asciidoc/guides/how-to-userinfo.adoc) (code [here](https://github.com/sjohnr/spring-authorization-server/tree/gh-537-how-to-customize-user-info/docs/src/docs/asciidoc/examples/src/main/java/sample/userinfo)).

> I'm not sure I see the distinction. The response is a 401, not a 403 or 404.

The 401 response is Spring Security's default response indicating that authentication is required, so that indeed is the distinction here. Once you provide a mechanism for authenticating the request, you will no longer receive that error. It is not exposed by default because we are seeking the smallest possible security footprint for the out-of-the-box solution. Exposing the userinfo endpoint is not risky *per se* but we want it to be an explicit taken by the application developer.

> Also, I'm not seeing anything in section and subsections of 5.3 of https://openid.net/specs/openid-connect-core-1_0.html that suggests that /userinfo is optional.

The userinfo response provides claims about the end-user. There are two ways to obtain those claims, the `id_token` and the `/userinfo` endpoint.

> This specification defines a set of standard Claims. They can be requested to be returned either in the UserInfo Response, per [Section 5.3.2](https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse), or in the ID Token, per [Section 2](https://openid.net/specs/openid-connect-core-1_0.html#IDToken).

The ID Token is the primary element of the OpenID Connect 1.0 specification:

> The primary extension that OpenID Connect makes to OAuth 2.0 to enable End-Users to be Authenticated is the ID Token data structure.

The UserInfo endpoint is not required to obtain these claims for authentication, and while not specifically stated appears optional for all intents and purposes. The OpenID Connect 1.0 specification(s) are written somewhat differently than OAuth 2.0/2.1, so that may account for the lack of clarity on this point.

If that interpretation is correct, the primary purpose of the UserInfo endpoint is not 100% clear from context alone, but likely provides an alternative to storing the ID Token on the client side, though it does not appear to be an alternative for validating the authentication response when using OIDC.

Does that make sense?

**Comment by aheitzmann** [comment]
Re: documentation, that's great to hear, and will be super helpful.

Re: the default 401 response, we may have to agree to disagree, but I'll try to clarify my position a bit more :)

In general, if there were a standard that defined an endpoint that could optionally be implemented by a service, and the spec for that endpoint stated that it must support bearer token authentication, but leaves discretion to the implementer as far as authorization rules are concerned, then I would argue that a service could chose to either
a) Not implement the optional endpoint, in which case callers should get a 404 back if they attempt to call it
b) Implement the optional endpoint, in which case bearer token authentication *must* be supported, meaning that a caller that passes a valid bearer token should not receive a 401, though depending on the authorization configured they could get a 403. 
In other words, from the perspective of a caller passing a valid bearer token, a 401 response indicates that the service both implements the optional endpoint (no 404), but doesn't support the authentication mechanism that the spec requires, which seems like a bug. 

WRT this project, I'd also note that the caller doesn't care whether their authorization server is implemented on top of spring-security or something else, so spring-security conventions aside, if you agree with the points above, I think the *default* behavior of the authorization server should be to either return 401 or 403 for /userinfo calls.

Re: UserInfo being optional, I'll admit I'm not an expert on OIDC or the intended use cases for the UserInfo endpoint, and I accept your position

**Comment by sjohnr** [comment]
Hello @aheitzmann, thanks for clarifying your point of view. It feels we are entering the territory that can only be expressed in terms of opinions and not interpretation of the spec. However, we’ll take your points under advisement so thank you for providing them!
